File: backend/apps/admin_api/serializers/analytics.py
from rest_framework import serializers


class SystemAttemptRowSerializer(serializers.Serializer):
    """One row of the real-data top-10-nursing-systems-by-attempts bar chart."""

    id = serializers.IntegerField()
    name = serializers.CharField()
    attempts = serializers.IntegerField()


class WeakSystemRowSerializer(serializers.Serializer):
    """One row of the real-data weakest-5-nursing-systems table."""

    id = serializers.IntegerField()
    name = serializers.CharField()
    attempts = serializers.IntegerField()
    correct = serializers.IntegerField()
    correct_rate = serializers.FloatField()


class RevenuePointSerializer(serializers.Serializer):
    month = serializers.CharField()
    revenue = serializers.IntegerField()


class RevenueSeriesSerializer(serializers.Serializer):
    points = RevenuePointSerializer(many=True)
    is_sample = serializers.BooleanField()


class SubscriptionMixPointSerializer(serializers.Serializer):
    tier = serializers.CharField()
    percentage = serializers.FloatField()


class SubscriptionMixSerializer(serializers.Serializer):
    points = SubscriptionMixPointSerializer(many=True)
    is_sample = serializers.BooleanField()


class AdminAnalyticsSerializer(serializers.Serializer):
    """
    Serializes the plain dict returned by
    apps.admin_api.services.analytics.build_admin_analytics().

    total_revenue is a DecimalField so it round-trips through JSON as a
    fixed-precision string (e.g. "0.00") rather than a float that could
    misrepresent cents. mom_student_growth and avg_quiz_score are
    allow_null=True because both are legitimately undefined on a database
    with no prior-month baseline / no completed quiz sessions yet — see
    services/analytics.py for why a fabricated number would be worse than
    null here.
    """

    total_students = serializers.IntegerField()
    total_revenue = serializers.DecimalField(max_digits=12, decimal_places=2)
    mom_student_growth = serializers.FloatField(allow_null=True)
    total_questions_answered = serializers.IntegerField()
    top_systems_by_attempts = SystemAttemptRowSerializer(many=True)
    avg_quiz_score = serializers.FloatField(allow_null=True)
    completion_rate = serializers.FloatField()
    weakest_systems = WeakSystemRowSerializer(many=True)
    revenue_series = RevenueSeriesSerializer()
    subscription_mix = SubscriptionMixSerializer()

File: frontend/src/types/analytics.ts
export interface SystemAttemptRow {
  id: number;
  name: string;
  attempts: number;
}

export interface WeakSystemRow {
  id: number;
  name: string;
  attempts: number;
  correct: number;
  correct_rate: number;
}

export interface RevenuePoint {
  month: string;
  revenue: number;
}

export interface RevenueSeries {
  points: RevenuePoint[];
  is_sample: boolean;
}

export interface SubscriptionMixPoint {
  tier: string;
  percentage: number;
}

export interface SubscriptionMix {
  points: SubscriptionMixPoint[];
  is_sample: boolean;
}

/** Mirrors AdminAnalyticsSerializer (backend: apps/admin_api/serializers/analytics.py). */
export interface AdminAnalytics {
  total_students: number;
  total_revenue: string;
  mom_student_growth: number | null;
  total_questions_answered: number;
  top_systems_by_attempts: SystemAttemptRow[];
  avg_quiz_score: number | null;
  completion_rate: number;
  weakest_systems: WeakSystemRow[];
  revenue_series: RevenueSeries;
  subscription_mix: SubscriptionMix;
}

export const ANALYTICS_TILE_LABELS = {
  totalStudents: "Total students",
  totalRevenue: "Total revenue",
  momGrowth: "Growth vs. last month",
  questionsAnswered: "Questions answered",
  avgQuizScore: "Average quiz score",
  completionRate: "Quiz completion rate",
} as const;

export const SAMPLE_DATA_BADGE_LABEL = "Sample data — live revenue will appear once subscriptions are active";

/**
 * Brand-neutral categorical palette for the two sample charts and the
 * top-systems bar chart. Kept small and local to this page rather than a
 * shared design-system export, since no other page in the app charts
 * anything yet.
 */
export const CHART_COLORS = ["#4338ca", "#d97706", "#16a34a", "#0ea5e9", "#dc2626"];
